src/coreclr/System.Private.CoreLib/src/System/RuntimeHandles.cs

```diff
@@ -46,12 +46,28 @@ private static RuntimeType GetRuntimeTypeFromHandleSlow(IntPtr handle)
             return typeObject!;
         }
 
-        [MethodImpl(MethodImplOptions.InternalCall)]
-        private static extern RuntimeType? GetRuntimeTypeFromHandleIfExists(IntPtr handle);
+        // implementation of CORINFO_HELP_GETSYNCFROMCLASSHANDLE, CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPE, CORINFO_HELP_TYPEHANDLE_TO_RUNTIMETYPEHANDLE
+        internal static unsafe RuntimeType GetRuntimeTypeFromHandle(IntPtr handle)
+        {
+            TypeHandle h = new((void*)handle);
+            if (h.IsTypeDesc)
+            {
+                return h.AsTypeDesc()->ExposedClassObject ?? GetRuntimeTypeFromHandleSlow(handle);
+            }
+
+            // The MethodTable fast path is in managed, so handle that here.
+            return GetRuntimeType(h.AsMethodTable());
```

Comment:
The slow path of this call is going to be `GetRuntimeTypeFromHandleSlow` as well. I am wondering whether the JIT smart enough to de-dup it with the TypeDesc slow path or whether it would be an improvement to dedup it manually.

Reply:
I’m fine collapsing them manually. I was being lazy and wanting to use an existing function. I think it probably makes more sense for it to be more self contained anyways.